p4-farm-nft/src/vault.rs

```diff
@@ -0,0 +1,293 @@
+        nft_contract_id: &NftContractId,
+        token: TokenId,
+    ) -> bool {
+        // find index for staking token into Contract.stake_tokens
+        
+        let nft_contract_i = find_acc_idx(nft_contract_id, &self.stake_nft_tokens);
+        let mut v = self.get_vault(&user);
+
+        // firstly check cheddar stake
+        let total_staked_tokens = v.get_number_of_staked_tokens();
+
+        // we expect for user who stake one more token have enough cheddar staked
+        let expected = expected_cheddar_stake(total_staked_tokens, self.cheddar_rate);
+        assert!(
+            v.total_cheddar_staked >= expected, 
+            "You need to stake {} yoctoCheddar more to stake one more NFT token",            
+            expected - v.total_cheddar_staked >= expected, 
+        );
+
+        // then update the past rewards
+        self.ping_all(&mut v);
+        // after that add "token" to staked into vault
+        v.staked[nft_contract_i].push(token.clone());
+        // update total staked info about this token
+        self.total_stake[nft_contract_i] += 1;
+
+        self._recompute_stake(&mut v);
+        self.vaults.insert(user, &v);
+        log!(
+            "Staked {}@{}, stake_units: {}", 
+            nft_contract_id, token.clone(), v.min_stake
+        );
+
+        true
+    }
+    /// Returns boost stake operation status.
+    /// Stake works only for 1 NFT token coming at the moment.
+    /// Revert transfer if nft_contract (`predecessor_account_id`) not in `Contract.boost_nft_contracts`
+    pub (crate) fn internal_boost_stake(
+        &mut self,
+        user: &AccountId,
+        nft_contract_id: &NftContractId,
+        token: TokenId
+    ) -> bool {
+        // find index for boost token into Contract.boost_nft_contracts
+        // Option to refund if `nft_contract_i` not in required for stake NFT contracts
+        let nft_contract_i = find_acc_idx(&nft_contract_id, &self.boost_nft_contracts);
+        let mut v = self.get_vault(&user);
+
+        if !v.boost_nft.is_empty() {
+            log!("Account already has boost NFT deposited. You can only deposit one");
+            return false
+        }
+        let nft_id: ContractNftTokenId = format!("{}{}{}", nft_contract_id, NFT_DELIMETER, token);
+        log!("Staking {} NFT - you will obtain a special farming boost", nft_id);
+        self.ping_all(&mut v);
+        v.boost_nft = nft_id.clone();
+
+        // update total staked info about this token
+        self.total_boost[nft_contract_i] += 1;
+
+        self._recompute_stake(&mut v);
+        self.vaults.insert(&user, &v);
+        log!(
+            "Added boost to user @{} with {}", 
+            user, contract_token_id.clone()
+        );
+        true
+    }
+
+    /// Returns remaining amount of NFTs from `nft_contract_id` which user has staked after function call.    
+    /// Panics if `token_id` is not supported or not staked by a user.
+    pub(crate) fn internal_nft_unstake(
```

Comment:
```suggestion
    pub(crate) fn _withdraw_nft(
```

Reply:
I don't think we need the `internal` prefix. It's not really popular outside of some near contracts.

Reply:
same for functions above